lib/gds-sso/authorised_user_constraint.rb

```diff
@@ -0,0 +1,20 @@
+module GDS
+  module SSO
+    class AuthorisedUserConstraint
```

Comment:
It'd be good to have an integration test for this.

I'd suggest adding this around a route in: https://github.com/alphagov/gds-sso/blob/main/spec/internal/config/routes.rb and adding a system test that checks a non logged in request, a request without necessary permission and one with necessary permission

Reply:
I'm trying this: [0aa3a1a](https://github.com/alphagov/gds-sso/pull/300/commits/0aa3a1aa5f5fb0ce405b8385c0c2247299892d73) but it is giving me hard time. Just posted the WIP changes in case you can see what i'm doing wrong while i'm debugging.

Seems like the constraint errors are not rescued by the controller_methods or the railtie, the system spec is raising the error without rescuing. The combustion gem is supposed to include the railtie as well, i tried to run the railtie manually too but no luck

Reply:
So Rails needs to be configured to handle the resuce responses as I understand.

If I change [`Combustion.initialize!`](https://github.com/alphagov/gds-sso/blob/84da55d0ce05197c695af776c8b26b04d68af09b/spec/spec_helper.rb#L9-L11) to be:

```
Combustion.initialize! :all do
  config.cache_store = :null_store
  config.action_dispatch.show_exceptions = true
end
```

I can get HTTP responses. However I'm getting a 500 rather than a 403 so just looking into why

Reply:
Yes this option I think renders the actual error page instead of stopping on raise. So I think you get the classic red  rails error page, saying: `GDS::SSO::PermissionDeniedError in ExampleController#constraint_restricted\nSorry, you don't seem to have the execute permission for this app.\n`. with 500 status code.

Reply:
I did also find the test works if we change Combustion::Initialize! to:

```
Combustion.initialize! :all do
  config.cache_store = :null_store
  config.action_dispatch.show_exceptions = true
  config.action_dispatch.rescue_responses.merge!(
    "GDS::SSO::PermissionDeniedError" => :forbidden
  )
end
```

so not sure our code is initialising in the right place. Does the 403 work in govuk-chat? i.e. is this a quirk of the test suite or a feature that isn't working.

Reply:
I also notice it works if we move the code out the initializer block: 

```
    class Railtie < Rails::Railtie
      initializer "gds-sso.initializer" do
        GDS::SSO.config do |config|
          config.cache = Rails.cache
          config.api_only = Rails.configuration.api_only
        end
        OmniAuth.config.logger = Rails.logger
      end

      config.action_dispatch.rescue_responses.merge!(
        "GDS::SSO::PermissionDeniedError" => :forbidden,
      )
    end
```

Though that still seems a bit wrong. Will do some checking in rails code

Reply:
Looking at the ActiveRecord Railtie it is configured as above:  https://github.com/rails/rails/blob/1d13a7c44d205a67cdd692bad56399e41a4eea44/activerecord/lib/active_record/railtie.rb#L23-L28

So I think I just gave you lousy advice in: https://github.com/alphagov/gds-sso/pull/300#discussion_r1575876567

Reply:
Thanks Kevin 🚀